nomos-core/chain-defs/src/mantle/tx.rs

```diff
@@ -101,14 +101,139 @@ impl From<SignedMantleTx> for MantleTx {
     }
 }
 
-#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
 pub struct SignedMantleTx {
     pub mantle_tx: MantleTx,
     // TODO: make this more efficient
     pub ops_proofs: Vec<Option<OpProof>>,
     pub ledger_tx_proof: ZkSignature,
 }
 
+#[derive(Debug, thiserror::Error, Clone, PartialEq, Eq)]
+pub enum VerificationError {
+    #[error("Invalid signature for operation at index {op_index}")]
+    InvalidSignature { op_index: usize },
+    #[error("Missing required proof for {op_type} operation at index {op_index}")]
+    MissingProof {
+        op_type: &'static str,
+        op_index: usize,
+    },
+    #[error("Incorrect proof type for {op_type} operation at index {op_index}")]
+    IncorrectProofType {
+        op_type: &'static str,
+        op_index: usize,
+    },
+}
+
+impl SignedMantleTx {
+    /// Create a new `SignedMantleTx` and verify that all required proofs are
+    /// present and valid.
+    ///
+    /// This enforces at construction time that:
+    /// - `ChannelBlob` operations have a valid Ed25519 signature from the
+    ///   declared signer
+    /// - `ChannelInscribe` operations have a valid Ed25519 signature from the
+    ///   declared signer
+    pub fn new(
+        mantle_tx: MantleTx,
+        ops_proofs: Vec<Option<OpProof>>,
+        ledger_tx_proof: ZkSignature,
+    ) -> Result<Self, VerificationError> {
+        let tx = Self {
+            mantle_tx,
+            ops_proofs,
+            ledger_tx_proof,
+        };
+        tx.verify_ops_proofs()?;
+        Ok(tx)
+    }
+
+    /// Create a `SignedMantleTx` without verifying proofs.
+    /// This should only be used for `GenesisTx` or in tests.
+    #[doc(hidden)]
+    #[cfg(any(test, debug_assertions))]
+    #[must_use]
+    pub const fn new_unverified(
+        mantle_tx: MantleTx,
+        ops_proofs: Vec<Option<OpProof>>,
+        ledger_tx_proof: ZkSignature,
+    ) -> Self {
+        Self {
+            mantle_tx,
+            ops_proofs,
+            ledger_tx_proof,
+        }
+    }
+
+    // TODO: might drop proofs after verification
```

Comment:
Yes, I think we can create a new type to enforce verification in other method signatures as per the comment above.

Reply:
Yeah, I meant that. I thought i already answered this :cry: